entity-framework/core/managing-schemas/migrations/operations.md

```diff
@@ -0,0 +1,143 @@
+ms.technology: entity-framework-core
+---
+Custom Migrations Operations
+============================
+The MigrationBuilder API allows you to perform many different kinds of operations during a migration, but it's far from
+exhaustive. However, the API is also extensible allowing you to define your own operations. There are two to extend the
+API: Using the `Sql()` method, or by defining custom `MigrationOperation` objects.
+
+To illustrate, let's look at implementing an operation that creates a database user using each approach. In our
+migrations, we want to enable writing the following code:
+
+``` csharp
+migrationBuilder.CreateUser("SQLUser1", "Password");
+```
+
+Using MigrationBuilder.Sql()
+----------------------------
+The easiest way to implement a custom operation is to define an extension method that calls `MigrationBuilder.Sql()`.
+Here is an example that generates the appropriate Transact-SQL.
+
+``` csharp
+static MigrationBuilder CreateUser(
+    this MigrationBuilder migrationBuilder,
+    string name,
+    string password)
+    => migrationBuilder.Sql($"CREATE USER {name} WITH PASSWORD '{password}';");
+```
+
+If your migrations need to support multiple database providers, you can use the `MigrationBuilder.ActiveProvider`
+property. Here's an example supporting both Microsoft SQL Server and PostreSQL.
+
+``` csharp
+static MigrationBuilder CreateUser(
+    this MigrationBuilder migrationBuilder,
+    string name,
+    string password)
+{
+    switch (migrationBuilder.ActiveProvider)
+    {
+        case "Npgsql.EntityFrameworkCore.PostgreSQL":
+            return migrationBuilder
+                .Sql($"CREATE USER {name} WITH PASSWORD '{password}';");
+
+        case "Microsoft.EntityFrameworkCore.SqlServer":
+            return migrationBuilder
+                .Sql($"CREATE USER {name} WITH PASSWORD = '{password}';");
+    }
+}
+```
+
+This approach only works if you know every provider where your custom operation will be applied.
+
+Using a MigrationOperation
+---------------------------
+To decouple the custom operation from the SQL, you can define your own `MigrationOperation` to represent it. The
+operation is then passed to the provider so it can determine the appropriate SQL to generate.
+
+``` csharp
+class CreateUserOperation : MigrationOperation
+{
+    public string Name { get; set; }
+    public string Password { get; set; }
+}
+```
+
+With this approach, the extension method just needs to add one of these operations to `MigrationBuilder.Operations`.
+
+``` csharp
+static MigrationBuilder CreateUser(
+    this MigrationBuilder migrationBuilder,
+    string name,
+    string password)
+{
+    migrationBuilder.Operations.Add(
+        new CreateUserOperation
+        {
+            Name = name,
+            Password = password
+        });
+
+    return migrationBuilder;
+}
+```
+
+Each provider can now decide what SQL to generate for this operation in their `IMigrationsSqlGenerator` service. Here is
```

Comment:
Maybe note here something like "this approach only works if every provider knows how to generate SQL for the operation." To align with the similar statement above about .Sql